commons/src/main/java/org/calrissian/accumulorecipes/commons/support/qfd/QfdHelper.java

```diff
@@ -196,61 +185,44 @@ public void save(Iterable<? extends T> items) {
-
-                            ColumnVisibility columnVisibility = colVisCache.get(visCache.getVis());
-                            if(columnVisibility == null) {
-                                columnVisibility = new ColumnVisibility(visCache.getVis());
-                                colVisCache.put(visCache.getVis(), columnVisibility);
-                            }
-
-                            Collection<Map<String, Object>> metas = tupleByObjectCache.get(visCache);
-                            metas = Collections2.transform(metas, removeVisFunction);
-
-                            value.set(metas.size() > 0 ? metadataSerDe.serialize(metas) : EMPTY_VALUE.get());
-
-                            String id = buildId(item);
-                            forwardCF.set(id);
-                            forwardCQ.set(key + NULL_BYTE + aliasValue);
-                            invertedCF.set(PREFIX_FI + NULL_BYTE + key);
-                            invertedCQ.set(aliasValue + NULL_BYTE + id);
-
-                            // forward mutation
-                            shardMutation.put(forwardCF,
-                                forwardCQ,
-                                columnVisibility,
-                                value);
-
-                            // reverse mutation
-                            shardMutation.put(invertedCF,
-                                invertedCQ,
-                                columnVisibility,
-                                value);  // forward mutation
-                        }
+                    for (Tuple tuple : item.getTuples()) {
+                        String visibility = getVisibility(tuple.getMetadata(), "");
+                        String aliasValue = typeRegistry.getAlias(tuple.getValue()) + ONE_BYTE +
+                            typeRegistry.encode(tuple.getValue());
+
+                        ColumnVisibility columnVisibility = new ColumnVisibility(visibility);
+
+                        Map<String,Object> meta = tuple.getMetadata();
+
+                        value.set(metadataSerDe.serialize(meta));
+
+                        forwardCF.set(id);
+                        forwardCQ.set(tuple.getKey() + NULL_BYTE + aliasValue);
+                        invertedCF.set(PREFIX_FI + NULL_BYTE + tuple.getKey());
+                        invertedCQ.set(aliasValue + NULL_BYTE + id);
+
+                        // forward mutation
+                        docKeys.add(new Key(new Text(shardId), forwardCF, forwardCQ, columnVisibility, 0));
+                        docValues.add(new Value(value));
+
+                        shardMutation.put(invertedCF,
+                            invertedCQ,
+                            columnVisibility,
+                            value);
                     }
+                    Value row = encodeRow(docKeys, docValues);
+                    shardMutation.put(new Text(id), new Text(), row);
```

Comment:
I have a fix for this on the accumulo side but it's not going in until 1.7.0 and we just agreed on new versioning semantics which would will pretty much guarantee that porting the change to a previous version will not be accepted. 

All that being said- the change allows the scan-time authorizations set on a ScannerBase to be propagated through the iterator stack w/ IteratorEnvironment so that we can apply the proper authorizations @ that time. Without this- unfortunately, we are not going to be able to guarantee that someone in the shell has the appropriate authorizations to view data. For now, maybe I'll try your second suggestion. Once that part works, an upgrade to 1.7.0 will not change much.